src/alignment/sparse.rs

```diff
@@ -0,0 +1,272 @@
+
+    let k = k as u32;
+
+    // incoming matches must be sorted to let us find the predecssor kmers by binary search.
+    for i in 1 .. matches.len() {
+        assert!(matches[i-1] < matches[i]);
+    }
+
+    let mut events: Vec<(u32, u32, u32)> = Vec::new();
+    let mut n = 0;
+
+    for (idx, &(x,y)) in matches.iter().enumerate() {
+        events.push((x,y,(idx+matches.len()) as u32));
+        events.push((x+k,y+k,idx as u32));
+
+        n = max(n, x+k);
+        n = max(n, y+k);
+    }
+    events.sort(); 
+
+
+    let mut max_col_dp: MaxBitTree<(u32, u32)> = MaxBitTree::new(n as usize);
+    let mut dp: Vec<(u32, i32)> = Vec::with_capacity(events.len());
+    let mut best_dp = (0,-1);
+
+    for _ in 0 .. events.len() {
+        dp.push((0,0));
+    }
+
+    for ev in events {
+        let p = (ev.2 % matches.len() as u32) as usize;
+        let j = ev.1;
+        let is_start = ev.2 >= (matches.len() as u32);
+        
+
+        if is_start {
+            dp[p] = (k, -1);
+            let (best_value, best_position) = max_col_dp.get(j as usize);
+            if best_value > 0 {
+                dp[p] = (k + best_value, best_position as i32);
+                best_dp = max(best_dp, (dp[p].0, p as i32));
+            }
+        } else {
+            // See if this kmer continues a diffent kmer
+            if ev.0 >= k+1 && ev.1 >= k+1 {
+                match matches.binary_search(&(ev.0-k-1, ev.1-k-1)) {
+                    Ok(cont_idx) => {
+                        let prev_score = dp[cont_idx].0;
+                        let candidate = (prev_score + 1, cont_idx as i32);
+                        dp[p] = max(dp[p], candidate);
+                        best_dp = max(best_dp, (dp[p].0, p as i32));
+                    },
+                    _ => (),
+                }
+            }
+
+            max_col_dp.set(ev.1 as usize, (dp[p].0, p as u32));
+        }
+    }
+
+    let mut traceback = Vec::new();
+    let (best_score, mut prev_match) = best_dp;
+    while prev_match >= 0 {
+        traceback.push(prev_match as usize);
+        prev_match = dp[prev_match as usize].1;
+    }
+    traceback.reverse();
+
+    (traceback, best_score, dp)
+}
+
+pub fn find_kmer_matches<T: AsRef<[u8]>>(seq1: &T, seq2: &T, k: usize) -> Vec<(u32, u32)> {
+
+    let slc1 = seq1.as_ref();
+    let slc2 = seq2.as_ref();
+
+    let mut set: HashMap<&[u8], Vec<u32>> = HashMap::new();
+    let mut matches = Vec::new();
+
+    for i in 0 .. slc1.len() - k + 1 {
+        set.entry(&slc1[i..i+k]).or_insert_with(|| Vec::new()).push(i as u32);
+    }
+
+    for i in 0 .. slc2.len() - k + 1 {
+        let slc = &slc2[i..i+k];
+        match set.get(slc) {
+            Some(matches1) => {
+                for pos1 in matches1 {
+                    matches.push((*pos1, i as u32));
+                }
+            },
+            None => (),
+        }
+    }
+
+    matches.sort();
+    matches
+}
+
+
+struct MaxBitTree<T: Default + Ord> {
```

Comment:
Can you move the MaxBitTree into a separate module within data structures? It sounds like a general purpose DS.